src/stack/array_stack.rs

```diff
@@ -0,0 +1,134 @@
+use crate::stack::Stack;
+
+/// Uses a Vector of a given capacity (16 by default)
+struct ArrayStack<T: Clone> {
+    array: Vec<Option<T>>,
+    head: Option<usize>,
+    size: usize,
+    capacity: usize,
+}
+
+impl<T: Clone> Stack<T> for ArrayStack<T> {
+    /// Pushes item of type `T` at the position pointed by `head`
+    /// Capacity will be updated once `capacity == count`
+    /// Amortized complexity: O(1)
+    fn push(&mut self, item: T) -> () {
+        if self.capacity() == self.count() {
+            self.update_capacity(self.capacity * 2);
+        }
+        let new_head = match self.head {
+            None => 0,
+            Some(h) => h + 1,
+        };
+        self.array[new_head] = Some(item);
+        self.head = Some(new_head);
+        self.size += 1;
+    }
+
+    /// Pops item of type `T` pointed by `head`
+    /// Sets `array[head]` to None
+    /// Capacity will be updated once `count` will become equals or less then 25% of the capacity
+    /// Amortized complexity: O(1)
+    /// Note that element needs to be cloned
+    fn pop(&mut self) -> T {
+        let curr_head = self.head_index();
+        let value = self.array[curr_head].clone();
+        let item = match value {
+            None => panic!("Cannot call pop. No value at the head of the stack."),
+            Some(item) => item,
+        };
+        self.array[curr_head] = None;
+        self.head = if curr_head == 0 {
+            None
+        } else {
+            Some(curr_head - 1)
+        };
+        self.size -= 1;
+        if self.count() <= self.capacity() / 4 {
+            self.update_capacity(self.capacity / 2);
+        }
+        item
+    }
+
+    fn peek(&self) -> T {
+        match self.array[self.size - 1].clone() {
+            None => panic!("Cannot call peek on an empty stack"),
+            Some(item) => item,
+        }
+    }
+
+    fn count(&self) -> usize {
```

Comment:
Since Rust seems to be happy with a backing field named as its read accessor (c.p. `capacity` property defined above), I would propose to aligh the name of the backing field to the name of the property, so that it's clear that they refer to the same underlying information.